appsrc/dcfile/dcarith.man
.TH DCARITH 1 "18 April 2008" "DICOM PS3" "DICOM PS3 - Arithmetic on DICOM pixels"
.SH NAME
dcarith \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Arithmetic on DICOM pixels
.SH SYNOPSIS
.HP 10
.B dcarith
[
.B \-v|verbose
]
[
.B \-ignorereaderrors
]
[
.B \-maskhighbits bits
]
[
.B \-addvalueoffset offset
]
[
.B \-zerovalue value
]
[
.B \-replacevalue startvalue endvalue newvalue
]
[
.B \-scale scalefactor
]
[
.B \-makeidentityrescale
]
.so man1/optin.so
.so man1/optout.so
.SH DESCRIPTION
.LP
.B dcarith
reads the named dicom or acr-nema input file and copies the information and
pixel data to a new dicom file, performing the same arithmetic computation on
all of the pixel data values (i.e., a point operation).
.LP
Note that the command only changes the stored pixel values in PixelData, and does not
affect any other attributes, including PixelPaddingValue, except as described
below for specific options. This means that other attributes that describe the
pixel data may need to be changed manually (using, for example, the \-r option).
.SH OPTIONS
The verbose output goes to standard error.
.LP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-verbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading, once read, during replacement, and before writing.
.RE
.TP
.B \-ignorereaderrors
.RS
Normally dccp stops if read errors are encountered, but if this option is
turned on, it will try to write what it can. This may be useful to remove
trailing garbage after a valid dicom set that has been padded past the
pixel data (e.g. to a blocksize) or similar.
.RE
.TP
.B \-pixelpaddingvalue
.RS
Apply the same operation to PixelPaddingValue as well as the stored pixel values. Implied by \-makeidentityrescale.
.RE
.TP
.B \-maskhighbits bits
.RS
Set to zero all bits above (to the left of) the specified bit.
.RE
.TP
.B \-addvalueoffset offset
.RS
Add a fixed offset value to all pixels.
.RE
.TP
.B \-zerovalue value
.RS
Sets all stored pixel values with the specified value to a new value of zero.
.RE
.TP
.B \-replacevalue startvalue endvalue newvalue
.RS
Sets all stored pixel values greater than or equal to the start value and less than or equal to the end value to the specified new value.
.RE
.TP
.B \-scale scalefactor
.RS
Multiply all stored pixel values by the specified scale factor (which may be a floating point value).
.RE
.TP
.B \-makeidentityrescale
.RS
Determine from the existing RescaleSlope and RescaleIntercept values what computation to apply to the stored
pixel values to make RescaleSlope 1 and RescaleIntercept 0, and apply that
computation to the stored pixel values, setting PixelRepresentation to 1 (signed) and BitsStored to 16, regardless of their previous
values. Also applies the same computation to PixelPaddingValue (i.e., implies \-pixelpaddingvalue).
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
\ 
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dcarith.cc
#include "attrmxls.h"
#include "attrothr.h"
#include "attrval.h"
#include "attrtype.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "elmconst.h"

class MaskHighBitFilter : public PointFilterIndependentOfOffset<Uint16,Uint16> {
private:
	Uint16 mask;
public:
	MaskHighBitFilter(unsigned lowbitstokeep)
			: PointFilterIndependentOfOffset<Uint16,Uint16>()
		{
			Assert(lowbitstokeep > 0);
			Assert(lowbitstokeep <= 16);
			mask = Uint16((Uint32(1)<<lowbitstokeep)-1u);
//cerr << "MaskHighBitFilter::MaskHighBitFilter(): lowbitstokeep=" << dec << lowbitstokeep << endl;
//cerr << "MaskHighBitFilter::MaskHighBitFilter(): mask=0x" << hex << mask << dec << endl;
		}

	Uint16 filter(Uint16 value)
		{
			return value&mask;
		}
};

class AddValueOffsetFilter : public PointFilterIndependentOfOffset<Uint16,Uint16> {
private:
	int offset;
public:
	AddValueOffsetFilter(int o)
			: PointFilterIndependentOfOffset<Uint16,Uint16>()
		{
			offset=o;
//cerr << "AddValueOffsetFilter::AddValueOffsetFilter(): offset=" << dec << offset << endl;
		}

	Uint16 filter(Uint16 value)
		{
			return (Uint16(value+offset));
		}
};

class ZeroValueFilter : public PointFilterIndependentOfOffset<Uint16,Uint16> {
private:
	Uint16 valuetozero;
public:
	ZeroValueFilter(int v)
			: PointFilterIndependentOfOffset<Uint16,Uint16>()
		{
			valuetozero=Uint16(v);
//cerr << "ZeroValueFilter::ZeroValueFilter(): v=" << dec << v << endl;
//cerr << "ZeroValueFilter::ZeroValueFilter(): valuetozero=0x" << hex << valuetozero << dec << endl;
		}

	Uint16 filter(Uint16 value)
		{
			return value==valuetozero ? Uint16(0) : value;
		}
};

class ReplaceValueFilter : public PointFilterIndependentOfOffset<Uint16,Uint16> {
private:
	Uint16 valuetostartreplace;
	Uint16 valuetoendreplace;
	Uint16 replacementvalue;
public:
	ReplaceValueFilter(int rs,int re,int v)
			: PointFilterIndependentOfOffset<Uint16,Uint16>()
		{
			valuetostartreplace=Uint16(rs);
			valuetoendreplace=Uint16(re);
			replacementvalue=Uint16(v);
		}

	Uint16 filter(Uint16 value)
		{
			return (value >= valuetostartreplace && value <= valuetoendreplace) ? replacementvalue : value;
		}
};

class ScaleValueFilter : public PointFilterIndependentOfOffset<Uint16,Uint16> {
private:
	float scalefactor;
public:
	ScaleValueFilter(float s)
			: PointFilterIndependentOfOffset<Uint16,Uint16>()
		{
			scalefactor=s;
		}

	Uint16 filter(Uint16 value)
		{
			return value=(Uint16)(scalefactor*(float)value);
		}
};

class MakeIdentityRescaleFilter : public PointFilterIndependentOfOffset<Uint16,Uint16> {
private:
	Uint16 maskInput;
	Uint16 maskOutput;
	float slope;
	float intercept;
public:
	MakeIdentityRescaleFilter(float rescaleslope,float rescaleintercept,unsigned inputBits,unsigned outputBits)
			: PointFilterIndependentOfOffset<Uint16,Uint16>()
		{
			Assert(inputBits > 8);
			Assert(inputBits <= 16);
			maskInput = Uint16((Uint32(1)<<inputBits)-1u);
			Assert(outputBits > 8);
			Assert(outputBits <= 16);
			maskOutput = Uint16((Uint32(1)<<outputBits)-1u);
			
			slope = rescaleslope;
			intercept = rescaleintercept;
		}

	Uint16 filter(Uint16 value)
		{
			// assumes input is always unsigned (doesn't sign extend beyond mask)
			float hu = (slope*(float)(value&maskInput)) + intercept;
			return value=(Uint16)(hu)&maskOutput;
		}
};

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);
	DicomOutputOptions 	dicom_output_options(options);

	bool bad = false;
	
	unsigned bits=0;
	bool operation_maskhighbits=options.get("maskhighbits",bits);

	int offset=0;
	bool operation_addvalueoffset=options.get("addvalueoffset",offset);

	int valuetozero=0;
	bool operation_zerovalue=options.get("zerovalue",valuetozero);

	int valuetostartreplace = 0;
	int valuetoendreplace = 0;
	int replacementvalue = 0;
	bool operation_replacevalue=false;
	{
		int replacevalueArgs[3];
		int n = options.get("replacevalue",replacevalueArgs,3);
		if (n > 0) {
			if (n == 3) {
				operation_replacevalue=true;
				valuetostartreplace = replacevalueArgs[0];
				valuetoendreplace = replacevalueArgs[1];
				replacementvalue = replacevalueArgs[2];
			}
			else {
				cerr << "replacevalue needs 3 arguments, not " << dec << n << endl;
				bad=true;
			}
		}
	}

	float scalefactor=1.0;
	bool operation_scale=options.get("scale",scalefactor);

	bool makeidentityrescale=options.get("makeidentityrescale");

	Assert(!(operation_maskhighbits && operation_addvalueoffset && operation_zerovalue && operation_replacevalue && operation_scale));

	bool verbose=options.get("verbose") || options.get("v");
	bool ignorereaderrors=options.get("ignorereaderrors");
	bool pixelpaddingvalue=options.get("pixelpaddingvalue");
	if (makeidentityrescale) {
		pixelpaddingvalue = true;
	}

	dicom_input_options.done();
	dicom_output_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);
	DicomOutputOpenerFromOptions output_opener(
		options,dicom_output_options.filename,cout);

	cerr << dicom_input_options.errors();
	cerr << dicom_output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!dicom_input_options.good()
	 || !dicom_output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options
	 || bad) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< dicom_output_options.usage()
			<< " [-v|-verbose]"
			<< " [-ignorereaderrors]"
			<< " [-pixelpaddingvalue]"
			<< " [-maskhighbits bits]"
			<< " [-addvalueoffset offset]"
			<< " [-zerovalue value]"
			<< " [-replacevalue startvalue endvalue newvalue]"
			<< " [-scale scalefactor]"
			<< " [-makeidentityrescale]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);
	DicomOutputStream dout(*(ostream *)output_opener,
		dicom_output_options.transfersyntaxuid,
		dicom_output_options.usemetaheader,
		dicom_output_options.useimplicitmetaheader,
		dicom_output_options.addtiff);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream  log(cerr);
	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	if (!list.good()) {
		if (!ignorereaderrors) {
			log << list.errors();
			success=false;
		}
		log << EMsgDC(DatasetReadFailed) << endl;
	}

	if (success) {
		PointFilterIndependentOfOffset<Uint16,Uint16> *filter = 0;
		if (operation_maskhighbits) {
			filter = new MaskHighBitFilter(bits);
		}
		else if (operation_addvalueoffset) {
			filter = new AddValueOffsetFilter(offset);
		}
		else if (operation_zerovalue) {
			filter = new ZeroValueFilter(valuetozero);
		}
		else if (operation_replacevalue) {
			filter = new ReplaceValueFilter(valuetostartreplace,valuetoendreplace,replacementvalue);
		}
		else if (operation_scale) {
			filter = new ScaleValueFilter(scalefactor);
		}
		else if (makeidentityrescale) {
			Uint16 vBitsStored = AttributeValue(list[TagFromName(BitsStored)]);
			float vRescaleSlope = AttributeValue(list[TagFromName(RescaleSlope)]);
			float vRescaleIntercept = AttributeValue(list[TagFromName(RescaleIntercept)]);
			Uint16 newBitsStored = 16;
			filter = new MakeIdentityRescaleFilter(vRescaleSlope,vRescaleIntercept,vBitsStored,newBitsStored);
			list -= TagFromName(RescaleSlope);
			list -= TagFromName(RescaleIntercept);
			list -= TagFromName(PixelRepresentation);
			list -= TagFromName(BitsStored);
			list -= TagFromName(HighBit);
			list += new DecimalStringAttribute(TagFromName(RescaleSlope),1.0);
			list += new DecimalStringAttribute(TagFromName(RescaleIntercept),0.0);
			list += new UnsignedShortAttribute(TagFromName(PixelRepresentation),1);
			list += new UnsignedShortAttribute(TagFromName(BitsStored),newBitsStored);
			list += new UnsignedShortAttribute(TagFromName(HighBit),newBitsStored-1);
			// NB. do not forget to setSuppressScalingOnBitDepthConversion() later, else any switch in bit depth will cause undesirable scaling
		}
		Assert(filter);
		
		if (pixelpaddingvalue) {
			Attribute *aPixelPaddingValue = list[TagFromName(PixelPaddingValue)];
			if (aPixelPaddingValue) {
				Uint16 vPixelPaddingValue = AttributeValue(aPixelPaddingValue);
				vPixelPaddingValue = filter->filter(vPixelPaddingValue);
				list -= TagFromName(PixelPaddingValue);
				list += new UnspecifiedShortAttribute(TagFromName(PixelPaddingValue),vPixelPaddingValue);
			}
		}

		Attribute *aPixelData = list[TagFromName(PixelData)];

		if (aPixelData) {
			OtherUnspecifiedLargeAttributeBase *oPixelData = 0;

			if (!aPixelData->isOtherData()) {
				log << EMsgDC(PixelDataIncorrectVR) << endl;
				success=false;
			}
			else {
				oPixelData = aPixelData->castToOtherData();
				Assert(oPixelData);
				oPixelData->insertPixelPointTransform(filter);
				oPixelData->setSuppressScalingOnBitDepthConversion(makeidentityrescale);
			}
		}
		else {
			log << EMsgDC(MissingAttribute)
			    << " - \"PixelData\""
			    << endl;
			success=false;
		}

		// Don't write list warnings yet ...
		// done in usualManagedAttributeListWrite ...

		if (!usualManagedAttributeListWrite(list,dout,
			dicom_output_options,log,verbose)) success=false;
	}

	return success ? 0 : 1;
}

	
